src/delta.c

```diff
@@ -38,86 +38,165 @@
-		return;
+	hash_assign(&f.hash[0], hash);
+	return true;
+}
+
+static void apply_one_delta(char *from_file, char *to_staged, char *delta_file, char *to_hash)
+{
+	int ret;
+
+	char *prefixed_from_file;
+	string_or_die(&prefixed_from_file, "%s/%s", path_prefix, from_file);
+
+	ret = apply_bsdiff_delta(prefixed_from_file, to_staged, delta_file);
+	if (ret) {
+		fprintf(stderr, "Couldn't use delta file %s: bsdiff failed\n", delta_file);
+		goto end;
 	}
 
-	if (file->deltapeer->is_deleted) {
-		return;
+	xattrs_copy(prefixed_from_file, to_staged);
+
+	/* TODO: Remove this and rely on hash verification only? */
+	if (xattrs_compare(prefixed_from_file, to_staged) != 0) {
+		fprintf(stderr, "Couldn't use delta file %s: failed to set xattrs\n", delta_file);
+		goto remove_and_fail;
 	}
 
-	/* check if the full file is there already, because if it is, don't do the delta */
-	string_or_die(&filename, "%s/staged/%s", state_dir, file->hash);
-	if (lstat(filename, &stat) == 0) {
-		free_string(&filename);
-		return;
+	char hash[SWUPD_HASH_LEN];
+	if (!compute_hash_from_file(to_staged, &hash[0])) {
+		fprintf(stderr, "Couldn't use delta file %s: hash calculation failed\n", delta_file);
+		goto remove_and_fail;
 	}
-	do_delta(filename, file);
-	free_string(&filename);
+	if (!hash_equal(hash, to_hash)) {
+		fprintf(stderr, "Couldn't use delta file %s: application resulted in wrong hash\n", delta_file);
+		goto remove_and_fail;
+	}
+
+	goto end;
+
+remove_and_fail:
+	(void)remove(to_staged);
+end:
+	free_string(&prefixed_from_file);
 }
 
-static void do_delta(char *filename, struct file *file)
+/* Check if the delta filename is well-formed, if so return true and fill the from/to
+ * buffers with the corresponding hashes. Return false otherwise. */
+static bool check_delta_filename(const char *delta_name, char *from, char *to)
 {
-	char *origin;
-	char *dir, *base, *tmp = NULL, *tmp2 = NULL;
-	char *deltafile = NULL;
-	int ret;
-	struct stat sb;
+	/* Delta files have the form [FROM_VERSION]-[TO_VERSION]-[FROM_HASH]-[TO_HASH]. */
+	const char *s = delta_name;
 
-	string_or_die(&deltafile, "%s/delta/%i-%i-%s-%s", state_dir,
-		      file->deltapeer->last_change, file->last_change, file->deltapeer->hash, file->hash);
-
-	ret = stat(deltafile, &sb);
-	if (ret != 0) {
-		free_string(&deltafile);
-		return;
+	s = index(s, '-');
+	if (!s) {
+		return false;
 	}
+	s++;
 
-	tmp = strdup(file->deltapeer->filename);
-	tmp2 = strdup(file->deltapeer->filename);
-
-	dir = dirname(tmp);
-	base = basename(tmp2);
+	s = index(s, '-');
+	if (!s) {
+		return false;
+	}
+	s++;
 
-	string_or_die(&origin, "%s/%s/%s", path_prefix, dir, base);
+	/* Note: SWUPD_HASH_LEN accounts for the nul-terminator after the hash. */
 
-	ret = apply_bsdiff_delta(origin, filename, deltafile);
-	if (ret) {
-		unlink_all_staged_content(file);
-		goto out;
+	/* Two hashes (without zero) plus one dash in the middle. */
+	if (strlen(s) != ((2 * (SWUPD_HASH_LEN - 1)) + 1)) {
```

Comment:
I still hate that we have to add a -1 EVERYWHERE we use this variable, if 99% of the use case is wanting it without a null terminator it should be the size we use in the #define IMO.

Reply:
I see quite a few references to `SWUPD_HASH_LEN` is this commit without the -1 though. In this function, it's needed often though, so maybe just define/use a variable set to `SWUPD_HASH_LEN - 1`?

Reply:
Rewrote the function with the extra variable, plus a few changes to code and comments.